finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,719 @@
+    pub fn new(cooldown: Duration) -> Self {
+        Discovery {
+            cooldown,
+            last_broadcast: HashMap::new(),
+            last_response: HashMap::new(),
+            requested_authentications: HashMap::new(),
+            next_query: NodeIndex(0),
+        }
+    }
+
+    fn should_broadcast(missing_authorities_num: usize, total_node_count: NodeCount) -> bool {
+        // If we are not sure we know of at least one honest node.
+        missing_authorities_num * 3 > 2 * total_node_count.0
+    }
+
+    /// Returns messages that should be sent as part of authority discovery at this moment.
+    pub fn discover_authorities(&mut self, handler: &SessionHandler) -> Vec<DiscoveryCommand> {
+        let missing_authorities = handler.missing_nodes();
+        if missing_authorities.is_empty() {
+            return Vec::new();
+        }
+        let node_count = handler.node_count();
+        let authentication = handler.authentication();
+        if Self::should_broadcast(missing_authorities.len(), node_count) {
+            // We know of fewer than 1/3 authorities, broadcast our authentication and hope others
+            // respond in kind.
+            vec![authentication_broadcast(authentication)]
+        } else {
+            // Attempt learning about more authorities from the ones you already know.
+            let mut result = Vec::new();
+            let mut target = self.next_query;
+            while result.len() < NODES_TO_QUERY {
+                if let Some(peer_id) = handler.peer_id(&target) {
+                    result.push(request(
+                        missing_authorities.clone(),
+                        authentication.clone(),
+                        peer_id,
+                    ));
+                }
+                target = NodeIndex((target.0 + 1) % node_count.0);
+            }
+            self.next_query = target;
+            result
+        }
+    }
+
+    /// Checks the authentication using the handler and returns the addresses we should be
+    /// connected to if the authentication is correct.
+    fn handle_authentication(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> Vec<Multiaddr> {
+        if !handler.handle_authentication(authentication.clone()) {
+            return Vec::new();
+        }
+        authentication.0.addresses()
+    }
+
+    fn should_rebroadcast(&self, node_id: &NodeIndex) -> bool {
+        match self.last_broadcast.get(node_id) {
+            Some(instant) => Instant::now() > *instant + self.cooldown,
+            None => true,
+        }
+    }
+
+    fn handle_broadcast(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> (Vec<Multiaddr>, Option<DiscoveryCommand>) {
+        let addresses = self.handle_authentication(authentication.clone(), handler);
+        if addresses.is_empty() {
+            return (addresses, None);
```

Comment:
in some sense I don't like the syntax, where you assign var, check if it isn't empty and return it, but don't have a good idea how to solve it.

What if addresses isn't empty, but only address there is our own? Or there is only address of message sender? I think here are some more cases

Reply:
Second comment applies also to other functions

Reply:
I've been wondering whether it's better to return `Vec::new()` here, very unsure. Do you think this would be clearer?

If someone sends us an authentication with addresses that don't actually resolve to them then they just wasted their time. Our own libp2p address cannot be there, since then the handler would not have accepted it. If there is our own address with another peer id then using it shouldn't break anything in libp2p, other than not being able to connect to the node who sent us this, so again just wasted time.

Reply:
Yes, I thought about it also and I think it would be better.

But such messages shouldn't also waste our time. But I understand if it is not a case here.

Reply:
Changed.